hkmc2/shared/src/main/scala/hkmc2/codegen/Block.scala

```diff
@@ -135,44 +135,22 @@ sealed abstract class Block extends Product with AutoLocated:
-          case _ => firstBlockRev -> firstTail
-      case TryBlock(sub, finallyDo, rest) => getReversedFlattenAndTrueTail(rest, TryBlock(sub, finallyDo, acc))
-      case Assign(lhs, rhs, rest) => getReversedFlattenAndTrueTail(rest, Assign(lhs, rhs, acc))
-      case a@AssignField(lhs, nme, rhs, rest) => getReversedFlattenAndTrueTail(rest, AssignField(lhs, nme, rhs, acc)(a.symbol))
-      case AssignDynField(lhs, fld, arrayIdx, rhs, rest) => getReversedFlattenAndTrueTail(rest, AssignDynField(lhs, fld, arrayIdx, rhs, acc))
-      case Define(defn, rest) => getReversedFlattenAndTrueTail(rest, Define(defn, acc))
-      case HandleBlock(lhs, res, par, args, cls, handlers, body, rest) => getReversedFlattenAndTrueTail(rest, HandleBlock(lhs, res, par, args, cls, handlers, body, acc))
-      case t: BlockTail => acc -> t
+  lazy val flatten: Block =
     
-    // reverse the Block returnned from the previous function,
-    // which does not contain `Begin` (except for the nested ones),
-    // and whose tail must be the dummy `End("for flatten only")`
-    def rev(b: Block, t: Block): Block = b match
-      case Match(scrut, arms, dflt, rest) => rev(rest, Match(scrut, arms, dflt, t))
-      case Label(label, body, rest) => rev(rest, Label(label, body, t))
-      case TryBlock(sub, finallyDo, rest) => rev(rest, TryBlock(sub, finallyDo, t))
-      case Assign(lhs, rhs, rest) => rev(rest, Assign(lhs, rhs, t))
-      case a@AssignField(lhs, nme, rhs, rest) => rev(rest, AssignField(lhs, nme, rhs, t)(a.symbol))
-      case AssignDynField(lhs, fld, arrayIdx, rhs, rest) => rev(rest, AssignDynField(lhs, fld, arrayIdx, rhs, t))
-      case Define(defn, rest) => rev(rest, Define(defn, t))
-      case HandleBlock(lhs, res, par, args, cls, handlers, body, rest) => rev(rest, HandleBlock(lhs, res, par, args, cls, handlers, body, t))
-      case End(msg) => t
-      case _: BlockTail => ??? // unreachable
-      case Begin(sub, rest) => ??? // unreachable
+    def flattenConcat(a: Block, b: => Block): Block = a match
+      case Match(scrut, arms, dflt, rest) => Match(scrut, arms, dflt, flattenConcat(rest, b))
+      case Label(label, body, rest) => Label(label, body, flattenConcat(rest, b))
+      case Begin(sub, rest) => flattenConcat(sub, flattenConcat(rest, b))
+      case TryBlock(sub, finallyDo, rest) => TryBlock(sub, finallyDo, flattenConcat(rest, b))
+      case Assign(lhs, rhs, rest) => Assign(lhs, rhs, flattenConcat(rest, b))
+      case a@AssignField(lhs, nme, rhs, rest) => AssignField(lhs, nme, rhs, flattenConcat(rest, b))(a.symbol)
+      case AssignDynField(lhs, fld, arrayIdx, rhs, rest) => AssignDynField(lhs, fld, arrayIdx, rhs, flattenConcat(rest, b))
+      case Define(defn, rest) => Define(defn, flattenConcat(rest, b))
+      case HandleBlock(lhs, res, par, args, cls, handlers, body, rest) => HandleBlock(lhs, res, par, args, cls, handlers, body, flattenConcat(rest, b))
+      case End(_) => b
```

Comment:
Here you're discarding the current `End`, which may have contained useful info, even when what you're replacing it with is `End("")` with no info. Better make `b` an `End => Block` function starting as the identity and pass `a` here.